src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -339,10 +340,27 @@ internal static SafeSslHandle AllocateSslHandle(SafeFreeSslCredentials credentia
                     // if server actually requests a certificate.
                     Ssl.SslSetClientCertCallback(sslHandle, 1);
                 }
-
-                if (sslAuthenticationOptions.IsServer && sslAuthenticationOptions.RemoteCertRequired)
+                else // sslAuthenticationOptions.IsServer
                 {
-                    Ssl.SslSetVerifyPeer(sslHandle);
+                    if (sslAuthenticationOptions.RemoteCertRequired)
+                    {
+                        Ssl.SslSetVerifyPeer(sslHandle);
+                    }
+
+                    if (sslAuthenticationOptions.CertificateContext?.Trust?._sendTrustInHandshake == true)
+                    {
+                        SslCertificateTrust trust = sslAuthenticationOptions.CertificateContext!.Trust!;
+                        X509Certificate2Collection certList = (trust._trustList ?? trust._store!.Certificates);
+
+                        Debug.Assert(certList != null, "certList != null");
+                        foreach (X509Certificate2 cert in certList)
+                        {
+                            if (!Ssl.SslAddClientCA(sslHandle, cert.Handle))
+                            {
+                                Debug.Fail("Failed to add issuer to trusted CA list.");
```

Comment:
I would probably also add trace message so we have some evidence for release builds.

Reply:
This should only happen when more than max. number of `X509_NAME` are pushed to the stack. I'd wager we are safe here.

I wanted to add a comment there but did not push that yet.